### PR TITLE
Add Pgpool TLS

### PR DESCRIPTION
**Pull request by MobarakHsn**

### CODE REVIEW

File: apis/kubedb/v1alpha2/pgpool_types.go [hunk 1]
```diff
@@ -111,6 +111,16 @@ type PgpoolSpec struct {
 	// +kubebuilder:default={name: "default"}
 	// +optional
 	PodPlacementPolicy *core.LocalObjectReference `json:"podPlacementPolicy,omitempty"`
+
+	// SSLMode for both standalone and clusters. [disable;allow;prefer;require;verify-ca;verify-full]
+	SSLMode PgpoolSSLMode `json:"sslMode,omitempty"`
+
+	// ClientAuthMode for sidecar or sharding. (default will be md5. [md5;scram;cert])
```
Comment (ArnobKumarSaha) [inline apis/kubedb/v1alpha2/pgpool_types.go:118]: Use marker
`// +kubebuilder:default=md5`

File: apis/kubedb/v1alpha2/pgpool_helpers.go [hunk 2]
```diff
@@ -272,6 +337,22 @@ func (p *Pgpool) SetDefaults() {
...
+		if p.Spec.ClientAuthMode == "" {
+			p.Spec.ClientAuthMode = PgpoolClientAuthModeMD5
+		}
+	} else {
+		if p.Spec.SSLMode == "" {
+			p.Spec.SSLMode = PgpoolSSLModeDisable
+		}
+		if p.Spec.ClientAuthMode == "" {
+			p.Spec.ClientAuthMode = PgpoolClientAuthModeMD5
+		}
```
Comment (ArnobKumarSaha) [inline apis/kubedb/v1alpha2/pgpool_helpers.go:353]: Remove clientAuthMode parts, as we are defaulting with marker now.

File: apis/kubedb/v1alpha2/pgpool_helpers.go [hunk 3]
```diff
@@ -203,6 +210,64 @@ func (p *Pgpool) ServiceLabels(alias ServiceAlias, extraLabels ...map[string]str
 	return p.offshootLabels(meta_util.OverwriteKeys(p.OffshootSelectors(), extraLabels...), svcTemplate.Labels)
 }
 
+func (p *Pgpool) GetSSLMODE(appBinding *appcat.AppBinding) (string, error) {
+	sslmodeString := appBinding.Spec.ClientConfig.Service.Query
+	if sslmodeString == "" {
+		return "disable", nil
```
Comment (ArnobKumarSaha) [inline apis/kubedb/v1alpha2/pgpool_helpers.go:216]: return PgpoolSSLModeDisable, nil

File: apis/kubedb/v1alpha2/pgpool_helpers.go [hunk 4]
```diff
@@ -69,6 +72,10 @@ func (p *Pgpool) ConfigSecretName() string {
 	return meta_util.NameWithSuffix(p.OffshootName(), "config")
 }
 
+func (p *Pgpool) TLSSecretName() string {
+	return meta_util.NameWithSuffix(p.OffshootName(), "tls-certs")
+}
```
Comment (ArnobKumarSaha) [inline apis/kubedb/v1alpha2/pgpool_helpers.go:77]: Why this ? `GetCertSecretName()` not enough ?

**Review by ArnobKumarSaha — CHANGES_REQUESTED** [review]

File: apis/kubedb/v1alpha2/pgpool_helpers.go [hunk 5]
```diff
@@ -203,6 +210,64 @@ func (p *Pgpool) ServiceLabels(alias ServiceAlias, extraLabels ...map[string]str
 	return p.offshootLabels(meta_util.OverwriteKeys(p.OffshootSelectors(), extraLabels...), svcTemplate.Labels)
 }
 
+func (p *Pgpool) GetSSLMODE(appBinding *appcat.AppBinding) (PgpoolSSLMode, error) {
+	sslmodeString := appBinding.Spec.ClientConfig.Service.Query
```
Comment (souravbiswassanto) [inline apis/kubedb/v1alpha2/pgpool_helpers.go:214]: Guess you should give nill check

**Review by souravbiswassanto — CHANGES_REQUESTED** [review]

File: apis/kubedb/v1alpha2/pgpool_helpers.go [hunk 6]
```diff
@@ -203,6 +210,67 @@ func (p *Pgpool) ServiceLabels(alias ServiceAlias, extraLabels ...map[string]str
...
+		return false, err
+	}
+	sslMode, err := p.GetSSLMODE(&apb)
+	if err != nil {
+		return false, err
+	}
+	if apb.Spec.TLSSecret != nil || len(apb.Spec.ClientConfig.CABundle) > 0 || sslMode == PgpoolSSLModeRequire || sslMode == PgpoolSSLModePrefer || sslMode == PgpoolSSLModeAllow || sslMode == PgpoolSSLModeVerifyCA || sslMode == PgpoolSSLModeVerifyFull {
+		return true, nil
+	}
+	return false, nil
```
Comment (ArnobKumarSaha) [inline apis/kubedb/v1alpha2/pgpool_helpers.go:245]: Just this should be enough
```
	return apb.Spec.TLSSecret != nil || len(apb.Spec.ClientConfig.CABundle) > 0 || sslMode != PgpoolSSLModeDisable, nil
```

File: apis/kubedb/v1alpha2/pgpool_helpers.go [hunk 7]
```diff
@@ -203,6 +210,67 @@ func (p *Pgpool) ServiceLabels(alias ServiceAlias, extraLabels ...map[string]str
...
+	if sslmodeString == "" {
+		return PgpoolSSLModeDisable, nil
+	}
+	temps := strings.Split(sslmodeString, "=")
+	if len(temps) != 2 {
+		return "", fmt.Errorf("the sslmode is not valid. please provide the valid template. the temlpate should be like this: sslmode=<your_desire_sslmode>")
+	}
+	sslMode := PgpoolSSLMode(strings.TrimSpace(temps[1]))
+	return sslMode, nil
+}
```
Comment (ArnobKumarSaha) [inline apis/kubedb/v1alpha2/pgpool_helpers.go:227]: `return PgpoolSSLMode(strings.TrimSpace(temps[1])), nil`

**Review by ArnobKumarSaha — CHANGES_REQUESTED** [review]

**Review by ArnobKumarSaha — APPROVED** [review]

**Review by souravbiswassanto — APPROVED** [review]

**Pull request merged**